src/server/gc_worker/gc_worker.rs

```diff
@@ -436,6 +470,124 @@ where
         Ok((handled_keys, wasted_keys))
     }
 
+    fn raw_gc_keys(
+        &mut self,
+        keys: Vec<Key>,
+        safe_point: TimeStamp,
+        regions_provider: Option<(u64, Arc<dyn RegionInfoProvider>)>,
+    ) -> Result<(usize, usize)> {
+        let range_start_key = keys.first().unwrap().clone().into_encoded();
+        let range_end_key = {
+            let mut k = keys
+                .last()
+                .unwrap()
+                .to_raw()
+                .map_err(|e| EngineError::Codec(e))?;
+            k.push(0);
+            Key::from_raw(&k).into_encoded()
+        };
+
+        let mut snapshot = self
+            .engine
+            .snapshot_on_kv_engine(&range_start_key, &range_end_key)?;
+
+        let mut raw_modifies = MvccRaw::new();
+        let mut keys = get_keys_in_regions(keys, regions_provider)?;
+
+        let mut gc_info = GcInfo::default();
+        let mut next_gc_key = keys.next();
+        while let Some(ref key) = next_gc_key {
+            if let Err(e) = self.raw_gc_key(
+                safe_point,
+                key,
+                &mut raw_modifies,
+                &mut snapshot,
+                &mut gc_info,
+            ) {
+                GC_KEY_FAILURES.inc();
+                error!(?e; "Raw GC meets failure"; "key" => %key,);
+                // Switch to the next key if meets failure.
+                gc_info.is_completed = true;
+            }
+
+            if gc_info.is_completed {
+                next_gc_key = keys.next();
+                gc_info = GcInfo::default();
+            } else {
+                // Flush writeBatch to engine.
+                Self::flush_raw_gc(raw_modifies, &self.limiter, &self.engine)?;
+                // After flush, reset raw_modifies.
+                raw_modifies = MvccRaw::new();
+            }
+        }
+
+        Self::flush_raw_gc(raw_modifies, &self.limiter, &self.engine)?;
+
+        // TODO: To implement the metrics.
+        Ok((0, 0))
+    }
+
+    fn raw_gc_key(
+        &mut self,
+        safe_point: TimeStamp,
+        key: &Key,
+        raw_modifies: &mut MvccRaw,
+        kv_snapshot: &mut <E as Engine>::Snap,
+        gc_info: &mut GcInfo,
+    ) -> Result<()> {
+        let start_key = key.clone().append_ts(safe_point.prev());
+        let mut cursor = CursorBuilder::new(kv_snapshot, CF_DEFAULT).build()?;
+
+        let mut statistics = CfStatistics::default();
+
+        cursor.seek(&start_key, &mut statistics)?;
+
+        let mut is_latest_version = true;
+
+        while cursor.valid()? {
+            let engine_slice = cursor.key(&mut statistics);
+            if !Key::is_user_key_eq(engine_slice, key.clone().as_encoded()) {
+                break;
+            }
+
+            // Skip the latest version before the safe point
+            if is_latest_version {
+                is_latest_version = false;
+                cursor.next(&mut statistics);
+                continue;
+            }
```

Comment:
How is *the latest delete/expired version before safepoint* cleaned up? It seems to be missing.